pkg/controllers/elasticsearch/rolebinding/controller.go

```diff
@@ -0,0 +1,78 @@
+package rolebinding
+
+import (
+	"fmt"
+	"reflect"
+
+	rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	rbaclisters "k8s.io/client-go/listers/rbac/v1beta1"
+	"k8s.io/client-go/tools/record"
+
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+)
+
+type Interface interface {
+	Sync(*v1alpha1.ElasticsearchCluster) error
+}
+
+type roleBindingControl struct {
+	kubeClient        kubernetes.Interface
+	roleBindingLister rbaclisters.RoleBindingLister
+
+	recorder record.EventRecorder
+}
+
+var _ Interface = &roleBindingControl{}
+
+func NewController(
+	kubeClient kubernetes.Interface,
+	roleBindingLister rbaclisters.RoleBindingLister,
+	recorder record.EventRecorder,
+) Interface {
+	return &roleBindingControl{
+		kubeClient:        kubeClient,
+		roleBindingLister: roleBindingLister,
+		recorder:          recorder,
+	}
+}
+
+func ownerCheck(
+	role *rbacv1beta1.RoleBinding,
+	cluster *v1alpha1.ElasticsearchCluster,
+) error {
+	if !metav1.IsControlledBy(role, cluster) {
+		ownerRef := metav1.GetControllerOf(role)
+		return fmt.Errorf(
+			"foreign owned Role: "+
+				"A Role with name '%s/%s' already exists, "+
+				"but it is controlled by '%v', not '%s/%s'",
+			role.Namespace, role.Name, ownerRef,
+			cluster.Namespace, cluster.Name,
+		)
+	}
+	return nil
+}
+
+func (e *roleBindingControl) Sync(c *v1alpha1.ElasticsearchCluster) error {
+	desiredRoleBinding := roleBindingForCluster(c)
+	existingRoleBinding, err := e.roleBindingLister.RoleBindings(desiredRoleBinding.Namespace).Get(desiredRoleBinding.Name)
+	if k8sErrors.IsNotFound(err) {
+		_, err := e.kubeClient.RbacV1beta1().RoleBindings(desiredRoleBinding.Namespace).Create(desiredRoleBinding)
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	if err := ownerCheck(existingRoleBinding, c); err != nil {
+		return err
+	}
+	// TODO: avoid using `reflect` here
+	if !reflect.DeepEqual(desiredRoleBinding.RoleRef, existingRoleBinding.RoleRef) || !reflect.DeepEqual(desiredRoleBinding.Subjects, existingRoleBinding.Subjects) {
+		_, err = e.kubeClient.RbacV1beta1().RoleBindings(desiredRoleBinding.Namespace).Update(desiredRoleBinding)
+		return err
+	}
+	return nil
+}
```

Comment:
❓ And this function seems almost identical to `RoleControl.Sync`. Can we turn this into a generic `createOrUpdate` ? Here or in a follow up.

Reply:
This isn't identical - we check different heuristics to determine whether an update is required (see L73 in this file). We'd also then need to lose the type safety in the client in order to do this.

I'd rather consume a few kb more space on disk to make this code easier to read (through the use of types, etc.).

If Go's type system wasn't so picky about interfaces, we could potentially pass the `e.roleBindingLister.RoleBindings(desiredRoleBinding.Namespace)` to some kind of a generic function so that it can call `Update`/`Get`/`Create`, but the type signature of those functions is `Update(*v1beta1.RoleBinding) (*v1beta1.RoleBinding, error)` and not `Update(runtime.Object) (runtime.Object, error)`, and despite *v1beta1.RoleBinding *implementing* `runtime.Object`, the two functions are not compatible.